R/bundle.R

```diff
@@ -729,30 +721,37 @@ createAppManifest <- function(appDir, appMode, contentCategory, hasParameters,
 
       # if description is NA, application dependency may not be installed
       if (is.na(info$description[1])) {
-        msg <- c(msg, paste0(capitalize(assetTypeName), " depends on package \"",
+        errorMessages <- c(errorMessages, paste0(capitalize(assetTypeName), " depends on package \"",
                              name, "\" but it is not installed. Please resolve ",
                              "before continuing."))
         next
       }
 
-      # validate package source (returns an error message if there is a problem)
-      msg <- c(msg, validatePackageSource(deps[i, ]))
+      # validate package source (returns a message if there is a problem)
+      packageMessages <- c(packageMessages, validatePackageSource(deps[i, ]))
 
       # good to go
       packages[[name]] <- info
     }
   }
-  if (length(missing_url_sources)) {
-    # Err when packages lack repository URL. We emit a warning about each package (see
-    # snapshotDependencies) before issuing an error with this resolution advice.
+  if (length(packageMessages)) {
+    # Advice to help resolve installed packages that are not available using the current set of
+    # configured repositories. Each package with a missing repository has already been printed (see
+    # snapshotDependencies).
+    #
+    # This situation used to trigger an error (halting deployment), but was softened because:
+    #   * CRAN-archived packages are not visible to our available.packages scanning.
+    #   * Source-installed packages may be available after a manual server-side installation.
     #
-    # It's possible we cannot find a repository URL for other reasons, including when folks locally
-    # build and install packages from source. An incorrectly configured "repos" option is almost
-    # always the cause.
-    msg <- c(msg, sprintf("Unable to determine the location for some packages. Packages must come from a package repository like CRAN or a source control system. Check that options('repos') refers to a package repository containing the needed package versions."))
+    # That said, an incorrectly configured "repos" option is almost always the cause.
+    packageMessages <- c(packageMessages,
+                         "Unable to determine the location for some packages. Packages should be installed from a package repository like CRAN or a source control system. Check that options('repos') refers to a package repository containing the needed package versions.")
+    warning(paste(formatUL(packageMessages, '\n*'), collapse = '\n'), call. = FALSE, immediate. = TRUE)
```

Comment:
One potential nit: in some cases, users might turn warnings into errors via `options(warn = 2)` or similar. We could use `message()` instead to write the message without worry that R might be configured to treat warnings as fatal. Is that worth changing?

```suggestion
    message(paste(formatUL(packageMessages, '\n*'), collapse = '\n'))
```

(we might already be warning in other contexts though)

Reply:
I'm OK with either option (warning or message). If folks change warnings into errors, that gives the option of hard-feedback that this PR is removing. That feels like a positive, no?

I don't feel strongly about this.